core/src/main/java/forklift/consumer/MessageRunnable.java

```diff
@@ -155,6 +155,17 @@ public void run() {
                     this.lifeCycle.call(ProcessStep.Complete, this);
```

Comment:
Do you want the complete step to run before the final acknowledgment of message receipt?

Reply:
Good catch - probably not. 

I don't think we'd want plugins to do things as if everything was successful and then have the acknowledgment fail (in case the plugin isn't idempotent wrt to the lifecycle of a message).

That precludes hooks from skipping the acknowledgment if something fails, but that use case doesn't really exist yet.

Reply:
Thought about this some more, and changed my mind.

It's non-obvious (to me) what should happen for `Error` and `Invalid` when acknowledgment fails.

It could maybe work like this:
```
Validating -> Invalid => DoAck()
Validating -> Processing -> Error => DoAck()
Validating -> Processing => DoAck() -> Complete
```

That's probably the "most useful" placement of acknowledgments, but it's also more complicated to understand.

I think it makes sense to say that since messages can be reprocessed, it's also possible for any of the lifecycle events to occur again for a message when that happens, and so plugins should be idempotent if duplicate processing is a concern.

Especially since failed acknowledgment should be very rare, and I know of don't know of need for a plugin to operate only in conditions where the acknowledgment has succeeded, I favor the simpler approach we have currently.

Reply:
I think that the Ack should happen after everything has completed, and the message is fully processed, which is the most common means for an ACK in all protocols I've been involved with.  Plus, the original code prior to Todd's changes had it there as well, so I believe that was Matt's intent.